### DISCUSSION TITLE
SqlCeTransaction.Commit(CommitMode.Immediate) is not committed immediately

### DISCUSSION THREAD

**Discussion by alexey-leonovich** [linq2db/linq2db#4685] [General]
### Describe your issue

I use custom class for a connection pool for database connections (under the hood it has a `Stack<SqlCeConnection>`, connection is created if stack is empty and is returned to stack on Dispose). So connections remain opened to increase performance.
I use transactions in several ways and recently asked [how to use them properly](https://github.com/linq2db/linq2db/discussions/4636). But looks like there is only one correct way (at least for SQL Server Compact 4.0) because with all other ways in some "critical" situations data is lost (even when using `SqlCeTransaction.Commit(CommitMode.Immediate`)). My application process could be sometimes suddenly killed (OS hard reboot, power supply fail, etc.) and it must keep data consistency even in that situations. 

### Steps to reproduce

My [sample repo](https://github.com/alexey-leonovich/ProcessKillTest) to run it as-is. To simplify example I emulate my connection pool by connection that remains open (and not Disposed) and emulate hard reset with `Process.Kill()`. Run application and press **w** to fill database with 5 rows. After commiting transaction process is killed. Run application again and press **r** - there are **0** rows in database. My experiments have shown that:

1. Among [all varians](https://github.com/linq2db/linq2db/discussions/4636) of using transactions only starting SqlCeTransaction before DataConnection creation and passing it into DataConnection constructor works (5 records are commited to database):
```c#
using (var transaction = connection.BeginTransaction())
using (var db = new DataConnection(SqlCeTools.GetDataProvider(), transaction))
```
2. Even `DataConnection.BeginTransaction()` results in data loss (maybe because `DataConnectionTrasaction.Commit()` under the hood calls `SqlCeTransaction.Commit()` that uses `CommitMode.Deferred` [by default](https://learn.microsoft.com/en-us/previous-versions/sql/compact/sql-server-compact-4.0/7tdf8adw(v=vs.100)#remarks)):
```c#
using (var db = new DataConnection(SqlCeTools.GetDataProvider(), connection))
using (var transaction = db.BeginTransaction())
```
3. There is no such problem for SQLite (SQL Server Compact 4.0 specifig bug? Or SQLite transaction simply commits faster? Or WAL helps to read data on second run?)

```c#
using LinqToDB.Mapping;
using LinqToDB;
using System;
using System.Data.SqlServerCe;
using System.Diagnostics;
using System.IO;
using LinqToDB.Data;
using LinqToDB.DataProvider.SqlCe;
using System.Linq;
using LinqToDB.DataProvider.SQLite;

namespace ProcessKillTest
{
    [Table(Name = "Payments")]
    public sealed class Payment
    {
        [Column(IsPrimaryKey = true)]
        [Column(DbType = "TEXT", Configuration = ProviderName.SQLite)]
        public Guid Id { get; set; }

        [Column(DbType = "nvarchar(255)", CanBeNull = true)]
        public string Comment { get; set; }

        [Column] public DateTime Date { get; set; }

        [Column(Name = "at_sum", DbType = "numeric(29,4)", CanBeNull = true, Configuration = ProviderName.SqlCe)]
        [Column(Name = "at_sum", DbType = "numeric(15,4)", CanBeNull = true, Configuration = ProviderName.SQLite)]
        public decimal Sum { get; set; }
    }

    internal class Program
    {
        private const string ConnectionStringFormat =
            "Persist Security Info = False; Data Source = {0}; File Mode = 'Read Write'; Max Database Size = 2048; SSCE:Default Lock Timeout=300000; Flush interval = 1";

        private const string fileName = "test.sdf";

        static void Main()
        {
            var connection = new SqlCeConnection(string.Format(ConnectionStringFormat, fileName));
            CreateDatabaseIfNeeded(connection);
            connection.Open();

            while (true)
            {
                Console.WriteLine("Input w [write to a database], r [check database record count] or q [exit]:");
                switch (Console.ReadKey().Key)
                {
                    case ConsoleKey.W:
                        WriteToDatabase(connection);
                        break;
                    case ConsoleKey.R:
                        CheckRecordsCount(connection);
                        break;
                    case ConsoleKey.Q:
                        return;
                        break;
                    default:
                        Console.WriteLine($"{Environment.NewLine}Wrong key!");
                        break;
                }
            }
        }


        private static void CreateDatabaseIfNeeded(SqlCeConnection connection)
        {
            using (var db = new DataConnection(SqlCeTools.GetDataProvider(), connection))
            {
                if (!File.Exists(fileName))
                {
                    SqlCeTools.CreateDatabase(fileName);
                    db.CreateTable<Payment>();
                }
            }
        }

        private static void WriteToDatabase(SqlCeConnection connection)
        {
            using (var db = new DataConnection(SqlCeTools.GetDataProvider(), connection))
            using (var transaction = connection.BeginTransaction())
            {
                try
                {
                    for (int i = 0; i < 5; i++)
                    {
                        var recordToAdd = new Payment 
                            { Id = Guid.NewGuid(), Date = DateTime.Now, Sum = 100.0m, Comment = "Test" };
                        db.Insert(recordToAdd);
                    }

                    transaction.Commit(CommitMode.Immediate);
                }
                catch (Exception e)
                {
                    transaction.Rollback();
                    Console.WriteLine(e);
                    throw;
                }
            }
            //if you remove Process.Kill() then CheckRecordsCount() returns 5 records
            Process.GetCurrentProcess().Kill();
        }


        private static void CheckRecordsCount(SqlCeConnection connection)
        {
            using (var db = new DataConnection(SQLiteTools.GetDataProvider(), connection))
            {
                var recCount = db.GetTable<Payment>().Count();
                Console.WriteLine($"{Environment.NewLine}Records count in database: {recCount}");
                if (recCount != 5)
                    Console.WriteLine("!!!Missing records!!!");
            }
        }
    }
}

```

### Environment details

`Linq To DB` version: *5.4.1*

Database (with version): *SQL Server Compact 4.0*

ADO.NET Provider (with version): *Microsoft.SqlServer.Compact 4.0.8876.1*

Operating system: *Windows 10 22H2 x64 Pro*

.NET Version: *.NET Framework 4.7.2*

**Comment by MaceWindu** [comment]
I'm not sure what you expect us to do. If SqlCe doesn't save data synchronously with `CommitMode.Immediate` - there is nothing we can do too

**Reply by MaceWindu** [reply]
I think you should retest it with this hotfix installed : https://support.microsoft.com/en-gb/topic/kb2960153-on-demand-hotfix-update-package-for-sql-server-compact-4-0-service-pack-1-40478d86-0bb7-f4ae-ac8f-a37e0034b361

**Reply by MaceWindu** [reply]
Could be downloaded here https://github.com/ErikEJ/SqlCeNetCore/tree/master/tools/4.0.8902 as looks like MS removed it

**Reply by MaceWindu** [reply]
I've made some tests and it looks like somehow it depends on connection state. If I dispose connection after commit - it works as expected. Probably you shouldn't try to implement own connection pool

**Reply by MaceWindu** [reply]
Another option which could help is to close "pooled" connections in AppDomain.CurrentDomain.UnhandledException handler. Will not work for killed process, but for app errors should help

**Comment by MaceWindu** [comment]
Looks like I found what the issue. It is related to how you create transaction.

When you do it like that:

```cs
using (var db = new DataConnection(SqlCeTools.GetDataProvider(), connection))
using (var transaction = connection.BeginTransaction())
```

linq2db doesn't attach transaction to command as it doesn't know anything about it and your inserts are flushed using flush interval.


YOu should change your code to this:

```cs
using (var transaction = connection.BeginTransaction())
using (var db = new DataConnection(SqlCeTools.GetDataProvider(), transaction))
```

**Reply by alexey-leonovich** [reply]
```
using (var transaction = connection.BeginTransaction())
using (var db = new DataConnection(SqlCeTools.GetDataProvider(), transaction))
```
is exactly what I've ended up with for now, thank you.
But it's strange that DataConnectionTransaction is not working. The following code also results in data loss:
```
using (var db = new DataConnection(SqlCeTools.GetDataProvider(), connection))
using (var transaction = db.BeginTransaction())
```
Suppose that `DataConnectionTrasaction.Commit()` under the hood calls `SqlCeTransaction.Commit()` that uses `CommitMode.Deferred` [by default](https://learn.microsoft.com/en-us/previous-versions/sql/compact/sql-server-compact-4.0/7tdf8adw(v=vs.100)#remarks)

**Reply by MaceWindu** [reply]
It's because we call ADO.NET Commit API. I've added feature request to support commit mode natively : https://github.com/linq2db/linq2db/issues/4686

**Reply by alexey-leonovich** [reply]
Many thanks for your help.
